atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java

```diff
@@ -0,0 +1,190 @@
+import java.util.SortedMap;
+import one.util.streamex.EntryStream;
+import one.util.streamex.StreamEx;
+
+public final class ClusterMetadataUtils {
+    private ClusterMetadataUtils() {
+        // util class
+    }
+
+    /**
+     * Returns a mapping of Node to token ranges its host contains, where every partition key in the specified
+     * list is present in one of the token ranges on one host.
+     *
+     * <p>For example there is a cluster that has keyspace K with a simple replication strategy and replication factor
+     * 2, the ring of the cluster has a range of (0, 3], and there are 3 nodes A, B, and C, where:
+     *
+     * <ul>
+     *   <li>A has replicas for K for token ranges (1, 2] and (2, 3]
+     *   <li>B has replicas for K for token ranges (2, 3] and (0, 1]
+     *   <li>C has replicas for K for token ranges (0, 1] and (1, 2]
+     * </ul>
+     *
+     * <p>The value returned by calling getTokenMapping([A, B, C], metadata, K, [0.25, 0.5, 2.5]) could be: <br>
+     * { A : set[ range(2, 3] ] B : set[ range(0, 1] ] }
+     *
+     * <p>Another possible return value is: <br>
+     * { B : set[ range(0, 1], range(2, 3] ] }
+     *
+     * <p>The node chosen for each token range can be any replica, so it is non-deterministic.
+     *
+     * @param nodeSet The Cassandra nodes whose replicas to check
+     * @param metadata The Datastax driver metadata from the Cassandra cluster
+     * @param keyspace The keyspace containing the partition keys we want to map
+     * @param partitionKeyTokens The Cassandra tokens for the partition keys we want to map
+     * @return Mapping of Node to token ranges its host contains, where every partition key in the specified
+     * list is present in one of the token ranges on one host
+     */
+    @SuppressWarnings("ReverseDnsLookup")
+    public static Map<InetSocketAddress, Set<TokenRange>> getTokenMapping(
+            Collection<InetSocketAddress> nodeSet, Metadata metadata, String keyspace, Set<Token> partitionKeyTokens) {
+
+        Map<String, InetSocketAddress> nodeMetadataHostMap = StreamEx.of(nodeSet)
+                .mapToEntry(InetSocketAddress::getHostName)
+                .invert()
+                .toMap();
+        Map<Host, List<TokenRange>> hostToTokenRangeMap =
+                getTokenMappingForPartitionKeys(metadata, keyspace, partitionKeyTokens);
+
+        return EntryStream.of(hostToTokenRangeMap)
+                .mapValues(list -> (Set<TokenRange>) new HashSet<>(list))
+                .mapKeys(host -> {
+                    String hostname = host.getEndPoint().resolve().getHostName();
+                    Preconditions.checkArgument(
+                            nodeMetadataHostMap.containsKey(hostname),
+                            "Did not find corresponding Node to run repair",
+                            SafeArg.of("hostname", hostname));
+                    return nodeMetadataHostMap.get(hostname);
+                })
+                .toMap();
```

Comment:
Hmm, I think I prefer extracting a helper method instead of the long sequence of mapKeys:
```
        return KeyedStream.stream(hostToTokenRangeMap)
                .mapKeys(host -> lookUpAddress(nodeMetadataHostMap, host))
                .map(List::stream)
                .map(stream -> stream.collect(Collectors.toSet()))
                .collectToMap();
```